### PR TITLE
Initial info on QE writers, readers and shared snapshot and fix for 1275

### PR DESCRIPTION
**Pull request by ashwinstar**
- Added initial description about QE writers, readers and shared snapshot. Lot more to gain knowledge on and updating the documents in the process.
- QE readers are not required to participate in any transaction activity like commit or abort. There counter part QE writer is responsible for performing the same. This commit is just fixing the over-sight caused by commit a453004, as anyways for QE reader case we will return much earlier in function and never ever reach here.

### CODE REVIEW

File: src/backend/utils/time/sharedsnapshot.c [hunk 1]
```diff
@@ -9,7 +9,141 @@
...
+ * So, the theme is to share private, not-yet-committed session transaction
+ * information with the QE Readers so the SegMate process group can all work
+ * on the transaction correctly. [We mostly think of QE Writers/Readers being
+ * on the segments. However, masters have special purpose QE Reader called the
+ * Entry DB Singleton. So, the SegMate module also works on the master.]
+ *
+ * Each shared snapshot is local only to the segment database. High level
+ * Writer gang member establishes a local transaction, aquires the slot in
+ * shared snapshot shmem space and copies the snapshot information into shared
+ * memory where the other qExecs that are segmates can find it. Following
```
Comment (asimrp) [inline src/backend/utils/time/sharedsnapshot.c]: Do you want to say "where the other qExecs that are \"readers\" can find it."?

File: src/backend/utils/time/sharedsnapshot.c [hunk 2]
```diff
@@ -9,7 +9,141 @@
...
+ * defines what it means to be "segmates."  The shared snapshot slot is
+ * identified by this unique session id. The unique session id is sent in from
+ * the QD as a GUC called "mpp_session_id". So the slot's field "slotid" will
+ * store the "mpp_session_id" that WRITER to the slot will use. Readers of the
+ * slot will find the correct slot by finding the one that has the slotid
+ * equal to their own mpp_session_id.
+ *
+ * Single Writer: Mechanism is simplified by introducing the restriction of
+ * only having a single qExec in a set of segmates capable of writing. Single
+ * WRITER qExec is the only qExec amonst all of its segmates that will ever
```
Comment (asimrp) [inline src/backend/utils/time/sharedsnapshot.c]: /amonst/amongst/

**Comment by hlinnaka** [comment]
Brilliant! I have a lot better understanding of all this now, after reading this!

**Comment by ashwinstar** [comment]
Glad its helpful. Will be adding more as I gain insights to fix 1276. Plus,
noting down lot of rough edges which should be altered to make this area
smoother/simpler.

File: src/backend/utils/time/sharedsnapshot.c [hunk 3]
```diff
@@ -9,7 +9,141 @@
...
+ * statement. All of the qExecs that belong to a particular user on a
+ * particular segment database need to have consistent visibility. Idea used
+ * is called "Shared Local Snapshot". Shared-memory data structure
+ * SharedSnapshotSlot shares session and transaction information among
+ * session's gang processes on a particular database instance. The processes
+ * are called a SegMate process group.
+ *
+ * A SegMate process group is a QE (Query Executor) Writer process and 0, 1 or
+ * more QE Reader processes. Greenplum needed to invent a SegMate sharing
+ * mechanism because in Postgress there is only 1 backend and most needed
```
Comment (danielgustafsson) [inline src/backend/utils/time/sharedsnapshot.c]: s/Postgress/Postgres/

**Comment by danielgustafsson** [comment]
Thanks a lot for this, really appreciate these detailed comments!

**Pull request merged**